sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignature.java

```diff
@@ -114,11 +114,116 @@ static Builder builder() {
   }
 
   /** A type of optional parameter of the {@link DoFn.ProcessElement} method. */
-  public enum Parameter {
-    BOUNDED_WINDOW,
-    INPUT_PROVIDER,
-    OUTPUT_RECEIVER,
-    RESTRICTION_TRACKER
+  public static abstract class Parameter {
+
+    public interface Visitor<ResultT> {
+      ResultT visit(BoundedWindowParameter p);
+      ResultT visit(InputProviderParameter p);
+      ResultT visit(OutputReceiverParameter p);
+      ResultT visit(RestrictionTrackerParameter p);
+      ResultT visit(StateParameter p);
+
+      /**
+       * A base class for a visitor with a default method for cases it is not interested in.
+       */
+      public static abstract class Defaults<ResultT> implements Visitor<ResultT> {
+
+        protected abstract ResultT visitDefault(Parameter p);
+
+        @Override
+        public ResultT visit(BoundedWindowParameter p) {
```

Comment:
Yes, that would be consistent with how Scala implements case classes, as I recall, and the path I took for the `Trigger` to `TriggerStateMachine` conversion.